network-formula/network/wicked/files/usr/local/sbin/ifsafesalt

```diff
@@ -0,0 +1,137 @@
+#!/bin/bash
+# ifup/ifdown wrapper script ensuring safe operation if called remotely through Salt
+# Copyright (C) 2023 Georg Pfuetzenreuter <mail+opensuse@georg-pfuetzenreuter.net>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+set -Cu
+
+action="${1?Cannot operate without an action}"
+interface="${2?Cannot operate without an interface}"
+
+base='/etc/sysconfig/network'
+base_backup="$base/salt-backup"
+
+if [ "$action" == 'up' -o "$action" == 'down' ]
+then
+	call="if$action"
+else
+	echo 'Invalid action'
+	exit 1
+fi
+
+if ! command -v "$call" >/dev/null
+then
+	echo "Unable to locate $call"
+	exit 1
+fi
+
+if [ ! -f "$base_backup/ifcfg-$interface" ]
+then
+	echo "Missing backup configuration for interface $interface, refusing to operate."
+	exit 1
+fi
+
+# Get IP addresses of the Salt minion and master
+read minion_ip master_ip < <(ss -HntA tcp dst :4505 | awk '{ split($4, con_out, ":"); split($5, con_in, ":"); print con_out[1] " " con_in[1] }')
+
+# Get network interface the minion is using to connect to the master
+out_interface="$(ip -br a sh | awk -v ip=$minion_ip '$0 ~ ip { print $1 }')"
+
+check() {
+	if ! ping -c3 -w5 -q "$master_ip" >/dev/null
+	then
+		return 1
+	fi
+	if ! timeout 20 salt-call -t15 --out quiet test.ping
+	then
+		return 1
+	fi
+}
+
+rollback() {
+	rollback=yes
+	cp -v "$base_backup/ifcfg-$interface" "$base/ifcfg-$interface"
+}
+
+run() {
+	timeout --preserve-status -k 60 30 "$call" "$interface"
+}
+
+backup() {
+	if command -v old >/dev/null
+	then
+		old "$base_backup/ifcfg-$interface"
+	fi
+	cp -v "$base/ifcfg-$interface" "$base_backup/ifcfg-$interface"
+}
+
+run_cycle() {
+	if run
+	then
+		if [ "$action" == 'down' ]
+		then
+			echo 'Operation successful.'
+			exit 0
+		fi
+		if check
+		then
+			if [ "$rollback" = 'yes' ]
+			then
+				echo 'Rollback successful.'
+			else
+				echo 'Operation and validation successful.'
+				backup
+			fi
+			exit 0
+		else
+			printf 'Operation failed'
+			if [ "$rollback" = 'yes' ]
```

Comment:
If $rollback is != yes, then the printf above will print a line without \n.

Most boring solution: change both printf to echo - which means you might get two lines instead of one, but don't have issues with a missing \n.

Reply:
I think if `rollback` is `!= yes`, the function "finishes" and it skips to the two printf's at the end of the file, no?

Reply:
Well, yes, indeed - technically you are right.

However, printing half of a line inside the function and half of it at a completely different place of the script is confusing, and makes it easy to introduce funny errors with future changes.

I'd recommend to let the function only print complete lines including \n.

Reply:
OK, fair point. :-)

Reply:
Refactored the output handling.